### PR TITLE
[Draft] Move query definitions into `rustc_query_system`

### PR DESCRIPTION
**Pull request by Julian-Wollersberger**
... and out of `rustc_middle`, to improve bootstrap times. (cc #65031)

The goal here is to allow future simplifications of the query system, and to make `rustc_middle` smaller, so that `x.py check` and `x.py build` is more parallel.

The important part are the first two commits:
The query definitions in the `rustc_queries!` proc macro generates several `macro_rules` macros that are invoked in different parts of `rustc_middle`. The change here is that one of them, `query_description_stream!`, now gets all the needed `use` statements as an input, rather than relying that they are defined at the top of the file. This makes it possible to move the query definitions into `rustc_query_system` although types like `HirId` or `DefId` aren't available there.

The rest of this PR demonstrates one simplification that this schema enables:
`DepKind` doesn't depend on the query parameter types that are not accessible in `rustc_query_system `, so one of the macros generated by `rustc_queries!` can be invoked in `rustc_query_system` directly to generate the `DepKind` enum. This enables removing (most of) the `DepKind` trait and thus removing the generic argument in a bunch of places. (Basically every `<D>` in #77871.) This doesn't fully work yet, because eg. `DepKind::debug` needs the global `ImplicitTyCxt`.

I guess this is an alternative approach to #70951, or maybe a first step to that. (My next todo is reading that PR.) CC rust-lang/compiler-team#277
Can I get your opinions on this approach, @cjgillot and maybe @eddyb and @oli-obk ? Is this a direction worth pursuing?

This PR is more a proof-of-concept and needs cleanup:
* Update docs to the new location for query definitions
* do `git move` on the query definitions file to not lose history
* rebase on #77871 
* make the refactoring of DepKind in a separate PR
* measure how long `rustc_queries!` takes, so we know the actual improvement here.

### CODE REVIEW

**Comment by jyn514** [comment]
> do git move on the query definitions file to not lose history

This will have no effect, `git move` is the same as deleting the old files and running `git add` on the new ones. GitHub is just bad at following renames.

File: compiler/rustc_macros/src/query.rs [hunk 1]
```diff
@@ -555,14 +558,24 @@ pub fn rustc_queries(input: TokenStream) -> TokenStream {
...
                 $($macro)*(#cached_queries);
             }
         }
 
-        #query_description_stream
+        #[macro_export]
+        macro_rules! query_description_stream {
+            // capture all needed `use` statements.
+            ($($uses:tt)*) => {
+                $($uses)*
```
Comment (cjgillot) [inline compiler/rustc_macros/src/query.rs:572]: You can surround thins expansion by an ad-hoc module to avoid polluting the namespace.

**Comment by cjgillot** [comment]
FWIW, when I created `rustc_query_system`, I wanted to have the `DepGraph` parametered by the `DepContext`, instead of the `DepKind`. This required to have `DepGraph<TyCtxt<'_>>` with a lifetime parameter, which caused lifetime variance problems. The `read_deps` and `with_deps` functions in `DepKind` trait are the middle ground I found at the time.

**Comment by Julian-Wollersberger** [comment]
I currently don't have much time to work on this, so I'll close it for now.
And university starts again next week, so I won't have time in the foreseeable future.

**Comment by Julian-Wollersberger** [comment]
> measure how long rustc_queries! takes, so we know the actual improvement here.

FWIW, the proc macro takes 350ms to execute.
(Measured with `eprintln!("Query proc macro took {:?}.", end_time.duration_since(start_time));`)
The total `x.py check` time is 2sec less, but that's the variance I normally get.

**Pull request closed without merging**